testnet/stacks-node/src/nakamoto_node.rs

```diff
@@ -0,0 +1,305 @@
+/// This struct is used to set up the node proper and launch the p2p thread and relayer thread.
+/// It is further used by the main thread to communicate with these two threads.
+pub struct StacksNode {
+    /// Atlas network configuration
+    pub atlas_config: AtlasConfig,
+    /// Global inter-thread communication handle
+    pub globals: Globals,
+    /// True if we're a miner
+    is_miner: bool,
+    /// handle to the p2p thread
+    pub p2p_thread_handle: JoinHandle<()>,
+    /// handle to the relayer thread
+    pub relayer_thread_handle: JoinHandle<()>,
+}
+
+/// Types of errors that can arise during Nakamoto StacksNode operation
+#[derive(Debug)]
+pub enum Error {
+    /// Can't find the block sortition snapshot for the chain tip
+    SnapshotNotFoundForChainTip,
+    /// The burnchain tip changed while this operation was in progress
+    BurnchainTipChanged,
+    /// Error while spawning a subordinate thread
+    SpawnError(std::io::Error),
+    /// Injected testing errors
+    FaultInjection,
+    /// This miner was elected, but another sortition occurred before mining started
+    MissedMiningOpportunity,
+    /// Attempted to mine while there was no active VRF key
+    NoVRFKeyActive,
+    /// The parent block or tenure could not be found
+    ParentNotFound,
+    /// Something unexpected happened (e.g., hash mismatches)
+    UnexpectedChainState,
+    /// A burnchain operation failed when submitting it to the burnchain
+    BurnchainSubmissionFailed,
+    /// A new parent has been discovered since mining started
+    NewParentDiscovered,
+    // The thread that we tried to send to has closed
+    ChannelClosed,
+}
+
+impl StacksNode {
+    /// This function sets the global var `GLOBAL_BURNCHAIN_SIGNER`.
+    ///
+    /// This variable is used for prometheus monitoring (which only
+    /// runs when the feature flag `monitoring_prom` is activated).
+    /// The address is set using the single-signature BTC address
+    /// associated with `keychain`'s public key. This address always
+    /// assumes Epoch-2.1 rules for the miner address: if the
+    /// node is configured for segwit, then the miner address generated
+    /// is a segwit address, otherwise it is a p2pkh.
+    ///
+    fn set_monitoring_miner_address(keychain: &Keychain, relayer_thread: &RelayerThread) {
+        let public_key = keychain.get_pub_key();
+        let miner_addr = relayer_thread
+            .bitcoin_controller
+            .get_miner_address(StacksEpochId::Epoch21, &public_key);
+        let miner_addr_str = addr2str(&miner_addr);
+        let _ = monitoring::set_burnchain_signer(BurnchainSigner(miner_addr_str)).map_err(|e| {
+            warn!("Failed to set global burnchain signer: {:?}", &e);
+            e
+        });
+    }
+
+    pub fn spawn(
+        runloop: &RunLoop,
+        globals: Globals,
+        // relay receiver endpoint for the p2p thread, so the relayer can feed it data to push
```

Comment:
Nit: Argument comments go in method blocks

Reply:
Will add a rustdoc for these arguments.